Feat/build deno package (#407434)
NixOS/nixpkgs: doc/languages-frameworks/javascript.section.md

})
```

### buildDenoPackage {#javascript-buildDenoPackage}

> `buildDenoPackage` allows you to package Deno-based projects in Nixpkgs without the use of an auto-generated dependencies file (as used in [node2nix](#javascript-node2nix)).

Reviewer: ```suggestion
`buildDenoPackage` allows you to package [Deno](https://deno.com/) projects in Nixpkgs without the use of an auto-generated dependencies file (as used in [node2nix](#javascript-node2nix)).
```

It would help readers unfamiliar with a particular corner of the IT ecosystem if random-looking strings like project names always had a link to their definition. Especially if they're short, because in that case web search can be completely useless.

It works by utilizing Deno's cache functionality -- creating a reproducible cache that contains the dependencies of a project, and pointing Deno to it.

#### buildDenoDeps {#javascript-buildDenoPackage-buildDenoDeps}

For every `buildDenoPackage`, first, a [fixed output derivation](https://nix.dev/manual/nix/2.18/language/advanced-attributes.html?highlight=fixed%20output#adv-attr-outputHash) is
created with all the dependencies mentioned in the `deno.lock`.
This works as follows:

> 1. They are installed using `deno install`.
> 2. All non-reproducible data is pruned.
> 3. The directories `.deno`, `node_modules` and `vendor` are copied to `$out`.
> 4. The output of the FOD is checked against the `denoDepsHash`.
> 5. The output is copied into the build of `buildDenoPackage`, which is not an FOD.
> 6. The dependencies are installed again using `deno install`, this time from the local cache only.

Reviewer: ```suggestion
1. They are installed using `deno install`.
1. All non-reproducible data is pruned.
1. The directories `.deno`, `node_modules` and `vendor` are copied to `$out`.
1. The output of the FOD is checked against the `denoDepsHash`.
1. The output is copied into the build of `buildDenoPackage`, which is not an FOD.
1. The dependencies are installed again using `deno install`, this time from the local cache only.
```

This allows re-shuffling lines without additional intervention

The `buildDenoDeps` derivation is in `passthru`, so it can be accessed from a `buildDenoPackage` derivation with `.denoDeps`

Related options:
* `denoDepsHash`: The output hash of the `buildDenoDeps` fixed output derivation.
* `denoInstallFlags`: Flags passed to `deno install`.
Defaults to `[ "--allow-scripts" "--frozen" ]` for `buildDenoDeps`
and `[ "--allow-scripts" "--frozen" "--cached-only" ]` for `buildDenoPackage`

#### Private registries {#javascript-buildDenoPackage-private-registries}
There are currently 2 options, which enable the use of private registries in a `buildDenoPackage` derivation.

> **`denoDepsImpureEnvVars`**:

Reviewer: Please use definition lists for argument lists

```ini
@<scope>:registry=https://<domain>/<path to private registry>
//<domain>/<path to private registry>:_authToken=${NPM_TOKEN}
```

:::

::: {.note}
Neither approach is ideal. For `buildNpmPackage`, there exists a third
option called `sourceOverrides`, which allows the user to inject Nix packages into
the output `node_modules` folder.
Since a Nix build implicitly uses the SSH keys of the machine,
this offers a third option to access private packages.
But this creates the requirement, that the imported package is packaged with nix first,
and that the source code can be retrieved with SSH.
This is possible for Deno, too, albeit it not
completely analogous to `buildNpmPackage`'s solution.
However, it has not been implemented yet.
:::

#### Compile to binary {#javascript-buildDenoPackage-compile-to-binary}

It's possible to compile a Deno project to a single binary using `deno compile`.
The binary will be named like the `.name` property in `deno.json`, if available,
or the `name` attribute of the derivation.

Related options:

* `hostPlatform`: The [host platform](#ssec-cross-platform-parameters) the binary is built for. Defaults to `builtins.currentSystem`.

> Supported are `"x86_64-darwin", "aarch64-darwin", "x86_64-linux", "aarch64-linux"`.

Reviewer: ```suggestion

  Supported values:
  - `"x86_64-darwin"
  - "aarch64-darwin"
  - "x86_64-linux"
  - "aarch64-linux"`
```

This is both easier to scan and to manipulate

Reviewer: Also please indent stuff inside the list item, and please check the rendering with `nix-shell doc --run devmode` to make sure looks the way you think it does. Markdown is the devil's language, it accepts all strings.

Author: > nix-shell doc --run devmode

Btw, I noticed that something is broken with the dev mode. All hyperlinks refer to an `index.html` file which does not exist in dev mode, instead the file is called `manual.html`.

Reviewer: Yeah there's an issue for that.

The binary is copied to `$out/bin` in the `installPhase`.

: _Default:_ `null`

: It's prefixed by `denoWorkspacePath`.

*`denortPackage`* (Derivation; optional)

: The package used as the Deno runtime, which is bundled with the JavaScript code to create the binary.

: _Default:_ `pkgs.denort`

: Don't use `pkgs.deno` for this, since that is the full Deno CLI, with all the development tooling.

: If you're cross compiling, this needs to be the `denort` of the `hostPlatform`.

::: {.note}
The binary will be dynamically linked and not executable on NixOS without [nix-ld](https://github.com/nix-community/nix-ld)
or [other methods](https://unix.stackexchange.com/questions/522822/different-methods-to-run-a-non-nixos-executable-on-nixos).

```nix
# configuration.nix
{
  config,
  lib,
  pkgs,
  ...
}:
{
  programs.nix-ld.enable = true;
  programs.nix-ld.libraries = with pkgs; [
    glibc
    gcc-unwrapped
  ];
}
```

:::

:::{.example}

##### example binary build {#javascript-buildDenoPackage-compile-to-binary-example}

```nix
{ buildDenoPackage, pkgs }:
buildDenoPackage {
  pname = "myPackage";
  version = "0.1.0";
  denoDepsHash = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA";

>   src = pkgs.nix-gitignore.gitignoreSource [ ] ./.;

Reviewer 2: can't we let `callPackage` pass `nix-gitignore`?
